File: tests/bench_matching.py
"""Benchmarks for pattern matching: substring, multi-pattern, regex, FST (pytest-benchmark)."""

import pytest

from kaos_nlp_core.matching import (
    FstSet,
    MultiPatternMatcher,
    RegexMatcher,
    substring_count,
    substring_find_all,
    substring_find_all_case_insensitive,
    substring_find_first,
)

# --- Substring search on War and Peace ---


@pytest.mark.benchmark(group="substring")
@pytest.mark.parametrize(
    "needle",
    ["Prince", "the", "Natasha Rostova", "NONEXISTENT_STRING_XYZ"],
    ids=lambda n: n[:20],
)
def test_substring_find_all(benchmark, war_and_peace_text, needle):
    benchmark(substring_find_all, war_and_peace_text, needle)


@pytest.mark.benchmark(group="substring")
def test_substring_count_the(benchmark, war_and_peace_text):
    benchmark(substring_count, war_and_peace_text, "the")


@pytest.mark.benchmark(group="substring")
def test_substring_find_first_prince(benchmark, war_and_peace_text):
    benchmark(substring_find_first, war_and_peace_text, "Prince")


@pytest.mark.benchmark(group="substring")
def test_substring_case_insensitive_prince(benchmark, war_and_peace_text):
    benchmark(substring_find_all_case_insensitive, war_and_peace_text, "prince")


# --- Multi-pattern (Aho-Corasick) on War and Peace ---


@pytest.mark.benchmark(group="multi_pattern")
def test_multi_pattern_4(benchmark, war_and_peace_text):
    m = MultiPatternMatcher(["Prince", "Princess", "Count", "Countess"])
    benchmark(m.find_all, war_and_peace_text)


@pytest.mark.benchmark(group="multi_pattern")
def test_multi_pattern_20(benchmark, war_and_peace_text):
    m = MultiPatternMatcher(
        [
            "the",
            "and",
            "was",
            "for",
            "that",
            "with",
            "his",
            "her",
            "from",
            "they",
            "have",
            "this",
            "been",
            "would",
            "could",
            "their",
            "which",
            "about",
            "other",
            "into",
        ]
    )
    benchmark(m.find_all, war_and_peace_text)


@pytest.mark.benchmark(group="multi_pattern")
def test_multi_pattern_20_count(benchmark, war_and_peace_text):
    m = MultiPatternMatcher(
        [
            "the",
            "and",
            "was",
            "for",
            "that",
            "with",
            "his",
            "her",
            "from",
            "they",
            "have",
            "this",
            "been",
            "would",
            "could",
            "their",
            "which",
            "about",
            "other",
            "into",
        ]
    )
    benchmark(m.count, war_and_peace_text)


# --- Regex on Shakespeare ---


@pytest.mark.benchmark(group="regex")
def test_regex_word_boundary_love(benchmark, shakespeare_text):
    r = RegexMatcher(r"\b[Ll]ove\b")
    benchmark(r.find_all, shakespeare_text)


@pytest.mark.benchmark(group="regex")
def test_regex_four_digit_numbers(benchmark, shakespeare_text):
    r = RegexMatcher(r"\b\d{4}\b")
    benchmark(r.find_all, shakespeare_text)


@pytest.mark.benchmark(group="regex")
def test_regex_capitalized_words(benchmark, shakespeare_text):
    r = RegexMatcher(r"\b[A-Z][a-z]{3,}\b")
    benchmark(r.find_all, shakespeare_text)


@pytest.mark.benchmark(group="regex")
def test_regex_character_names(benchmark, shakespeare_text):
    r = RegexMatcher(r"\b(Romeo|Juliet|Hamlet|Othello|Macbeth|Lear)\b")
    benchmark(r.find_all, shakespeare_text)


# --- FST on War and Peace vocabulary ---


@pytest.fixture(scope="module")
def war_peace_vocab(war_and_peace_text):
    """Sorted unique words from War and Peace."""
    from conftest import DEFAULT_TOKENIZER

    return sorted(set(DEFAULT_TOKENIZER.tokenize_words(war_and_peace_text)))


@pytest.mark.benchmark(group="fst")
def test_fst_build(benchmark, war_peace_vocab):
    benchmark(FstSet, war_peace_vocab)


@pytest.mark.benchmark(group="fst")
def test_fst_contains(benchmark, war_peace_vocab):
    fst = FstSet(war_peace_vocab)
    queries = ["prince", "war", "peace", "love", "death", "nonexistent"]
    benchmark(lambda: [fst.contains(q) for q in queries])


@pytest.mark.benchmark(group="fst")
def test_fst_prefix_search(benchmark, war_peace_vocab):
    fst = FstSet(war_peace_vocab)
    benchmark(fst.prefix_search, "pr")


@pytest.mark.benchmark(group="fst")
def test_fst_fuzzy_d1(benchmark, war_peace_vocab):
    fst = FstSet(war_peace_vocab)
    benchmark(fst.fuzzy_search, "princ", 1)


@pytest.mark.benchmark(group="fst")
def test_fst_fuzzy_d2(benchmark, war_peace_vocab):
    fst = FstSet(war_peace_vocab)
    benchmark(fst.fuzzy_search, "princ", 2)

File: benches/bench_matching.rs
//! Benchmarks for pattern matching: substring, multi-pattern, regex, FST.

use criterion::{criterion_group, criterion_main, BenchmarkId, Criterion};
use std::hint::black_box;

use _rust::core::matching::{
    fst_match::FstSet,
    multi_pattern::{MultiPatternMatchKind, MultiPatternMatcher},
    regex_match::RegexMatcher,
    substring,
};

fn load_text(filename: &str) -> String {
    std::fs::read_to_string(format!("tests/fixtures/{filename}"))
        .unwrap_or_else(|_| "The quick brown fox jumps over the lazy dog. ".repeat(10000))
}

fn bench_substring_search(c: &mut Criterion) {
    let text = load_text("war_and_peace.txt");
    let mut group = c.benchmark_group("substring");

    for needle in ["Prince", "the", "Natasha Rostova", "NONEXISTENT_STRING_XYZ"] {
        group.bench_with_input(BenchmarkId::new("find_all", needle), needle, |b, needle| {
            b.iter(|| black_box(substring::find_all(&text, needle)))
        });
    }

    group.bench_function("count/the", |b| {
        b.iter(|| black_box(substring::count(&text, "the")))
    });

    group.bench_function("find_first/Prince", |b| {
        b.iter(|| black_box(substring::find_first(&text, "Prince")))
    });

    group.bench_function("case_insensitive/prince", |b| {
        b.iter(|| black_box(substring::find_all_case_insensitive(&text, "prince")))
    });

    group.finish();
}

fn bench_multi_pattern(c: &mut Criterion) {
    let text = load_text("war_and_peace.txt");
    let mut group = c.benchmark_group("multi_pattern");

    // Small pattern set
    let small = MultiPatternMatcher::new(
        &["Prince", "Princess", "Count", "Countess"],
        MultiPatternMatchKind::LeftmostFirst,
    )
    .unwrap();

    group.bench_function("4_patterns", |b| {
        b.iter(|| black_box(small.find_all(&text)))
    });

    // Medium pattern set — common English words
    let medium_patterns: Vec<&str> = vec![
        "the", "and", "was", "for", "that", "with", "his", "her", "from", "they", "have", "this",
        "been", "would", "could", "their", "which", "about", "other", "into",
    ];
    let medium =
        MultiPatternMatcher::new(&medium_patterns, MultiPatternMatchKind::LeftmostFirst).unwrap();

    group.bench_function("20_patterns", |b| {
        b.iter(|| black_box(medium.find_all(&text)))
    });

    group.bench_function("20_patterns/count", |b| {
        b.iter(|| black_box(medium.count(&text)))
    });

    group.finish();
}

fn bench_regex(c: &mut Criterion) {
    let text = load_text("shakespeare.txt");
    let mut group = c.benchmark_group("regex");

    // Simple word boundary match
    let word_re = RegexMatcher::new(r"\b[Ll]ove\b").unwrap();
    group.bench_function("word_boundary/love", |b| {
        b.iter(|| black_box(word_re.find_all(&text)))
    });

    // Date-like patterns
    let date_re = RegexMatcher::new(r"\b\d{4}\b").unwrap();
    group.bench_function("four_digit_numbers", |b| {
        b.iter(|| black_box(date_re.find_all(&text)))
    });

    // Capitalized words
    let cap_re = RegexMatcher::new(r"\b[A-Z][a-z]{3,}\b").unwrap();
    group.bench_function("capitalized_words", |b| {
        b.iter(|| black_box(cap_re.find_all(&text)))
    });

    // Complex alternation
    let alt_re = RegexMatcher::new(r"\b(Romeo|Juliet|Hamlet|Othello|Macbeth|Lear)\b").unwrap();
    group.bench_function("character_names", |b| {
        b.iter(|| black_box(alt_re.find_all(&text)))
    });

    group.finish();
}

fn bench_fst(c: &mut Criterion) {
    let text = load_text("war_and_peace.txt");
    let mut group = c.benchmark_group("fst");

    // Build vocabulary from the text
    let words: Vec<String> = text
        .split_whitespace()
        .map(|w| {
            w.trim_matches(|c: char| !c.is_alphanumeric())
                .to_lowercase()
        })
        .filter(|w| !w.is_empty())
        .collect();
    let unique: std::collections::BTreeSet<&str> = words.iter().map(|s| s.as_str()).collect();
    let vocab: Vec<&str> = unique.into_iter().collect();

    // Build FST from War and Peace vocabulary
    group.bench_function("build/war_and_peace_vocab", |b| {
        b.iter(|| black_box(FstSet::build(vocab.iter().copied()).unwrap()))
    });

    let fst = FstSet::build(vocab.iter().copied()).unwrap();
    let vocab_size = vocab.len();

    group.bench_function(format!("contains/{vocab_size}_terms"), |b| {
        b.iter(|| {
            for word in ["prince", "war", "peace", "love", "death", "nonexistent"] {
                black_box(fst.contains(word));
            }
        })
    });

    group.bench_function("prefix_search/pr", |b| {
        b.iter(|| black_box(fst.prefix_search("pr")))
    });

    group.bench_function("fuzzy_search/princ/d1", |b| {
        b.iter(|| black_box(fst.fuzzy_search("princ", 1).unwrap()))
    });

    group.bench_function("fuzzy_search/princ/d2", |b| {
        b.iter(|| black_box(fst.fuzzy_search("princ", 2).unwrap()))
    });

    group.finish();
}

criterion_group!(
    benches,
    bench_substring_search,
    bench_multi_pattern,
    bench_regex,
    bench_fst,
);
criterion_main!(benches);
